### PR TITLE
Add client.submissions.edit to edit an existing submission

### LINKED ISSUE #8: Add client.submissions.edit to edit an existing submission
## Use cases
Building block for a data cleaning pipeline that treats Central as the source of truth. User realizes one systematic error was made, they script identifying all of those errors and confirming them, then building updated submissions and submitting them using this method. They also specify a comment that looks something like "Scripted fix for fields foo, bar, baz".

Some kinds of issues I've fixed using similar building blocks:
- bad character encoding in CSV attachment (so you have F�� in your submissions)
- systematic place name misspellings (open field for administrative region, you want to normalize "Banaadir", "Bandir" to "Banadir")
- error in select underlying value that was fixed in later form version (so submissions from early form version have ff for female instead of f).

## Proposed inputs

`instance_id`: str
`form_id: str`
`comment: str`
`xml: str` - a string representation of the new XML to submit. Must have the correct `deprecatedID` and a new `instanceID`

`project_id: Optional[int] = None`

## Proposed output
Exception in case of any error state.

None in case of success. It could be an object representing the submission metadata after the update but I expect this would generally be ignored.

### PR DESCRIPTION
**Pull request by lindsay-stevens**
Closes #8 

- add client.submissions.edit to update submission data and optionally add a comment at the same time.
  - added Submission post, put, patch, edit
  - added Comment post, list
- validation improvements:
  - use pydantic validators instead of writing the same. Also benefits by returning a type-cast value if it was possible to do so, e.g. "1" is a castable number so return 1.
  - instead of checking 200 response status only, use the built-in `raise_for_status`, which checks for status codes >=400 and <600.

#### What has been done to verify that this works as intended?

Tests that return type is as expected when feeding in mock test fixture data. 

Note that currently the Central sandbox returns unexpected Comment data in response to POST. It includes an empty `actorId`, and the comment body text is split into an array. So until that is fixed, `comment.post` and `submission.edit` (if a comment is sent) will raise a validation error when processing the response data.

#### Why is this the best possible solution? Were any other approaches considered?

Implementation is as discussed / specified in #8. In terms of leaving the validation issue there, presumably this will be fixed up shortly. In the meantime users could catch the error and ignore it. This seems preferable to not validating the response for now, and remembering to go back and add it again once the central issue is fixed.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Only as described above.

#### Do we need any specific form for testing your changes? If so, please attach one.

No, there's test / mock fixture data.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.

The README is updated with the new methods. I also added in the first line from each method in case the verb / method names alone weren't clear about what they do.

#### Before submitting this PR, please make sure you have:
- [x] included test cases for core behavior and edge cases in `tests`
- [x] run `nosetests` and verified all tests pass
- [x] run `python bin/pre_commit.py` to format / lint code
- [x] verified that any code or assets from external sources are properly credited in comments

### CODE REVIEW

File: pyodk/endpoints/submissions.py [hunk 1]
```diff
@@ -159,12 +153,165 @@ def get_table(
...
+
+        response = self.session.response_or_error(
+            method="GET",
+            url=self.urls.get_table.format(project_id=pid, form_id=fid, table_name=table),
+            logger=log,
+            params=params,
+        )
+        return response.json()
+
+    def post(
```
Comment (lognaturel) [inline pyodk/endpoints/submissions.py:165]: In the spirit of matching informal/api docs language for naming library methods, how about `create`? This matches the header at https://odkcentral.docs.apiary.io/#reference/submissions/submissions/creating-a-submission When I had previously commented on `read`, I didn't want to deter from using all CRUD words! Especially now that we have the generic HTTP methods, I think "how do people talk about this thing" is a good guiding question.

File: pyodk/endpoints/submissions.py [hunk 2]
```diff
@@ -159,12 +153,165 @@ def get_table(
...
+            url=self.urls.post.format(project_id=pid, form_id=fid),
+            logger=log,
+            headers={"Content-Type": "application/xml"},
+            params=params,
+            data=xml,
+        )
+        data = response.json()
+        return Submission(**data)
+
+    def put(
```
Comment (lognaturel) [inline pyodk/endpoints/submissions.py:211]: How about making this one "private" (prepend with _ and don't document). In general we want to encourage users to make a comment with an edit. They can omit the `comment` parameter to `edit` if they really want. (I had suggested making it required! But I'm ok with it being optional)

Here's the Central frontend UI if you make an edit without a comment:
Screen Shot 2022-10-11 at 2 22 02 PM

We would have made it required if we could have but because of the way we communicate with Enketo we can't.

File: pyodk/endpoints/submissions.py [hunk 3]
```diff
@@ -159,12 +153,165 @@ def get_table(
...
+            method="PUT",
+            url=self.urls.put.format(project_id=pid, form_id=fid, instance_id=iid),
+            logger=log,
+            headers={"Content-Type": "application/xml"},
+            data=xml,
+        )
+        data = response.json()
+        return Submission(**data)
+
+    def patch(
```
Comment (lognaturel) [inline pyodk/endpoints/submissions.py:255]: `review`? `update_review_state`? I think even if the `patch` verb does eventually allow for updating more metadata we'll want to keep the ability to just update the review state.

For convenience, this method should also take a comment string parameter. Another option would be to make this method "private" and expose a separate `review`. That would match what I've proposed for `edit`/`put`.

There are some nice workflows that can be automated with this. For example, I've seen scripts that identify outlier submissions and change their review state to "has issues" with a useful comment. Then someone can use the frontend to review issues and act on them.

File: README.md [hunk 4]
```diff
@@ -96,15 +97,22 @@ The `Client` is specific to a configuration and cache file. These approximately
...
-  - get
-  - list
-  - get_table
+  - list: Read all Submission metadata.
+  - get: Read Submission metadata.
+  - get_table: Read Submission data.
+  - post: Create a Submission.
+  - put: Update Submission data.
+  - patch: Update Submission metadata.
+  - edit: Submission.post then Comment.post.
```
Comment (lognaturel) [inline README.md:112]: Edit a submission and optionally comment on it.

File: pyodk/endpoints/comments.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+class URLs(bases.Model):
+    class Config:
+        frozen = True
+
+    list: str = "projects/{project_id}/forms/{form_id}/submissions/{instance_id}/comments"
+    post: str = "projects/{project_id}/forms/{form_id}/submissions/{instance_id}/comments"
+
+
+class CommentService(bases.Service):
```
Comment (lognaturel) [inline pyodk/endpoints/comments.py:27]: There's a broader question around how to structure access to nested resources or resources that are not as central to Central as `projects`, `forms`, `submission`. This intersects a bit with the questions I raised around users at https://github.com/getodk/pyodk/issues/7 

In my mind a goal is to match user expectations and common usage over strictly mapping to the Central API. I think where we do match Central we want to match the frontend user experience more than the backend API.

Comments feel very much connected to submissions. What do you think of `client.submission.list_comments`, `client.submission.add_comment`? I realize this means a looser internal structure and maybe moving away from the `endpoints` package name.

File: pyodk/endpoints/submissions.py [hunk 6]
```diff
@@ -159,12 +153,165 @@ def get_table(
...
+            url=self.urls.get_table.format(project_id=pid, form_id=fid, table_name=table),
+            logger=log,
+            params=params,
+        )
+        return response.json()
+
+    def post(
+        self,
+        xml: str,
+        form_id: Optional[str] = None,
```
Comment (lognaturel) [inline pyodk/endpoints/submissions.py:168]: I'm surprised to see `form_id` optional here and on other signatures here. Was this maybe left over from the manager-based chaining?

**Review by lognaturel — CHANGES_REQUESTED** [review]
The `edit` method looks great. I like the validation, error checking and types. 👍

I have some questions and suggestions around some of the other methods introduced. I'm really glad we have the generic HTTP verb methods and I think that for our designed methods we should try to move away from those and take the opportunity to be more descriptive.

One quick option would be to make the other methods`_` prefixed and undocumented for now as we continue iterating on some of our broader goals and gathering more user feedback. My high-level sense is that a small number of carefully-designed and named methods are more likely to get used than an extensive 1-1 mapping with the backend API.

File: pyodk/endpoints/submissions.py (hunk 3 above)
Comment (lindsay-stevens) [reply pyodk/endpoints/submissions.py:255]: I'm OK with the name as-is but please let me know if you'd like to change it. Please confirm if you'd like a `comment` parameter in Submission.patch? Presumably this means review state changes accept a comment in the UI.

File: README.md (hunk 4 above)
Comment (lindsay-stevens) [reply README.md:112]: Did you want the above comment to replace all 4 highlighted lines? Or just like `edit: Edit a submission and optionally comment on it.`

File: pyodk/endpoints/comments.py (hunk 5 above)
Comment (lindsay-stevens) [reply pyodk/endpoints/comments.py:27]: I'm OK with the name/structure as-is but please let me know if you'd like to change it. There could be a higher level module/package that holds all these convenience methods, rather than mixing them into the endpoint mappings.

File: pyodk/endpoints/submissions.py (hunk 6 above)
Comment (lindsay-stevens) [reply pyodk/endpoints/submissions.py:168]: More or less, yeah. It's a similar idea as project_id, where you could set a default on the Service class instance, then that'd be used as a default for any calls using that instance. The parameter validation checks that either the default or an input is set. Please let me know what you'd like to change if anything.

**Comment by lognaturel** [comment]
@lindsay-stevens and I discussed and the next step is to only document and expose `edit` while keeping the infrastructure private.

File: pyodk/endpoints/submissions.py (hunk 1 above)
Comment (lindsay-stevens) [reply pyodk/endpoints/submissions.py]: Renamed to "create"

File: pyodk/endpoints/submissions.py (hunk 2 above)
Comment (lindsay-stevens) [reply pyodk/endpoints/submissions.py]: _put marked private

File: pyodk/endpoints/submissions.py (hunk 3 above)
Comment (lindsay-stevens) [reply pyodk/endpoints/submissions.py]: _patch marked private and `review` method added with comment param

File: README.md (hunk 4 above)
Comment (lindsay-stevens) [reply README.md]: README reworded per above method name changes

File: pyodk/endpoints/comments.py (hunk 5 above)
Comment (lindsay-stevens) [reply pyodk/endpoints/comments.py]: Marked most of pyodk as private (package name prefixed with underscore, and in top-level `__init__.py` specified public elements in `__all__`). Exceptions are `client.Client` and `errors` module. On `client.Client`, the `_comments` property is marked private. All the docs/examples only deal with the `Client`.

Added `list_comments` and `add_comment` methods to `Submission`. These just delegate to the existing `Comments` module. So given the above, the internal structure can be clean but what is exposed is as described above.

File: pyodk/endpoints/submissions.py (hunk 6 above)
Comment (lindsay-stevens) [reply pyodk/endpoints/submissions.py]: When we discussed this item, @lognaturel was going to consider whether this was useful to keep in or not. So for now it is still in.

**Review by lognaturel — APPROVED** [review]

File: pyodk/endpoints/submissions.py (hunk 6 above)
Comment (lognaturel) [reply pyodk/endpoints/submissions.py]: I have no strong feelings about this so let's leave it in.

**Comment by lognaturel** [comment]
Thanks so much, @lindsay-stevens.
